### PR TITLE
assemble all db repos

### LINKED ISSUE #54: Refactor our fragile cross-repo dependency setup
Comments like https://github.com/mozilla/fxa-auth-server/pull/954 and https://github.com/mozilla/fxa-auth-db-mysql/issues/51#issuecomment-110996075 make me sad.  The current factoring of these auth-db-* repos seems to be creating a lot of busywork and deployment footguns.  Let's find a better way to do this.

I personally lean towards "squash them all back together into a single repo" like the way oauth-server does its backends, but am open to other suggestions :-)

### PR DESCRIPTION
**Pull request by dannycoates**
fixes #54

New layout:

The db-server dep is moved in at the root level and required as a dependency by package.json and the mem bits are now intermixed in bin and lib. This allows everything to be in sync while allowing us to publish fxa-auth-db-server as a module to npm for 3rd party implementations. 

Eventually we might want to rename this repo to reflect that its not just mysql, but its no big deal to leave it, and nothing will need to change for fxa-dev or prod. I've already got this running locally with fxa-auth-server with only a few minor changes.

```
├── CHANGELOG
├── Dockerfile
├── Gruntfile.js
├── LICENSE
├── README.md
├── bin
│   ├── db_patcher.js
│   ├── mem.js
│   └── server.js
├── config
│   ├── config.js
│   └── index.js
├── coverage.html
├── fxa-auth-db-server
│   ├── AUTHORS
│   ├── CHANGELOG
│   ├── CONTRIBUTING.md
│   ├── Gruntfile.js
│   ├── LICENSE
│   ├── README.md
│   ├── coverage.html
│   ├── docs
│   ├── grunttasks
│   ├── index.js
│   ├── lib
│   ├── package.json
│   ├── scripts
│   └── test
├── index.js
├── lib
│   ├── db
│   ├── logging.js
│   ├── notifier.js
│   └── promise.js
├── npm-shrinkwrap.json
├── package.json
```

Note, this requires npm >= 2.0 for the "file" type dependency.

### CODE REVIEW

**Comment by jrgm** [comment]
Looks good. But looks like you'll need to `npm install -g npm@2` because the 0.10 build is using npm @1.4.28 unfortunately.

**Comment by dannycoates** [comment]
Thanks @jrgm, now we're green

**Comment by dannycoates** [comment]
maybe it would be less weird if instead of the npm dep if we just changed the `require`s. I was dazzled by the new shiny 'file:' option.

**Comment by dannycoates** [comment]
> maybe it would be less weird

except then we'd need to promote the db-server deps up. clearly i need to stop looking at this. :beer:

**Comment by chilts** [comment]
Things I've looked at which seem fine:

* `npm test` in the top level seems fine (including patching the database)
* `npm install && npm test` in fxa-auth-db-server seems fine
* `node bin/mem.js` seems fine
* `npm start` starts a mysql server fine

Random minor queries:

* `db_patcher.js` is now executable but is only ever run using node (doesn't have a  `#!` interpreter) ... minor, but is it needed?
* I think the tests are running the mysql backend (test/backend/db_tests.js requires '../../lib/db/mysql' same in `test/backend/remote.js`) but neither are doing the same with the mem backend, is that true? If so, should it do both? This is probably my main concern here. :cactus: 

Apart from delving into all of the code (which just looks like a re-arrangement with probable changes to requires) then this all looks fine to me! :) r+ with nit regarding backend tests. Totally bogus @dannycoates!

bogus

File: fxa-auth-db-server/.travis.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+language: node_js
```
Comment (rfk) [inline fxa-auth-db-server/.travis.yml]: Will this .travis.yml in a random subdirectory actually get run?

**Comment by rfk** [comment]
> I think the tests are running the mysql backend (test/backend/db_tests.js requires '../../lib/db/mysql'
> same in test/backend/remote.js) but neither are doing the same with the mem backend,

If true, we need to figure out a way to run both sets of tests under travis

**Comment by rfk** [comment]
IIUC the next step here is for danny to common chilts' queries above, then we can rebase for a merge.

File: fxa-auth-db-server/.travis.yml (hunk 1 above)
Comment (dannycoates) [inline fxa-auth-db-server/.travis.yml]: no

**Comment by rfk** [comment]
I went ahead and rebased this and merged in train-42 from the other repos.  Let's get it merged and we can file follow-up bugs for further cleanup. Merging r=@chilts

**Comment by dannycoates** [comment]
Eh, sorry I haven't looped back to this yet. @rfk, thanks for the merge anyway. I see @philbooth took the followups, lmk if I can help with anything :)

**Pull request merged**